### PR TITLE
Updated multiplayer character to not block on dynamic networked rigid…

### PR DESCRIPTION
**Pull request by SergeyAMZN**
… bodies

Signed-off-by: pereslav <pereslav@amazon.com>

### CODE REVIEW

File: Gem/Code/Source/Components/CharacterComponent.cpp [hunk 1]
```diff
@@ -10,9 +10,73 @@
...
+        // non-kinematic dynamic bodies should not impede the movement of the character
+        if (actor->getConcreteType() == physx::PxConcreteType::eRIGID_DYNAMIC)
+        {
+            const physx::PxRigidDynamic* rigidDynamic = static_cast<const physx::PxRigidDynamic*>(actor);
+
+            bool isKinematic = (rigidDynamic->getRigidBodyFlags() & physx::PxRigidBodyFlag::eKINEMATIC);
+            if (isKinematic)
+            {
+                const PhysX::ActorData* actorData = PhysX::Utils::GetUserData(rigidDynamic);
+                const AZ::EntityId entityId = actorData->GetEntityId();
```
Comment (AMZN-Olex) [inline Gem/Code/Source/Components/CharacterComponent.cpp:55]: Is actorData guaranteed to be non-null?

**Review by AMZN-Olex — APPROVED** [review]

File: Gem/Code/Source/Components/CharacterComponent.cpp (hunk 1 above)
Comment (SergeyAMZN) [reply Gem/Code/Source/Components/CharacterComponent.cpp]: good call! It won't be null if created via PhysX gem, but technically game code can create rigid dynamics directly with no user data (or incompatible user data), so added a null check here. thanks!

File: Gem/Code/Source/Components/CharacterComponent.cpp [hunk 2]
```diff
@@ -10,9 +10,76 @@
...
+
+            bool isKinematic = (rigidDynamic->getRigidBodyFlags() & physx::PxRigidBodyFlag::eKINEMATIC);
+            if (isKinematic)
+            {
+                const PhysX::ActorData* actorData = PhysX::Utils::GetUserData(rigidDynamic);
+                if (actorData)
+                {
+                    const AZ::EntityId entityId = actorData->GetEntityId();
+
+                    if (NetworkRigidBodyRequestBus::FindFirstHandler(entityId) != nullptr)
```
Comment (kberg-amzn) [inline Gem/Code/Source/Components/CharacterComponent.cpp:59]: You also could just try constructing a network entity handle.. if that succeeds then it's a net bound entity, if it doesn't then it's not. Also we should make sure the entity specifically has a net rigid body component on it, and not a net character component. We should also probably make net character incompatible with net rigid body so we can't put both on the same entity.

File: Gem/Code/Source/Components/CharacterComponent.h [hunk 3]
```diff
@@ -19,6 +20,7 @@ namespace MultiplayerSample
 {
     class CharacterComponent
```
Comment (kberg-amzn) [inline Gem/Code/Source/Components/CharacterComponent.h:21]: Just a heads up that I've renamed this component to NetworkCharacterComponent upon the advice of the UX team. AnimatedHitVolumes is also now NetworkHitVolumes, so aside from the camera and WASD movement all of our multiplayer sample components are prefixed with 'Network'. It should merge cleanly but it's good to let you know this is coming.

File: Gem/Code/Source/Components/CharacterComponent.cpp [hunk 4]
```diff
@@ -41,6 +108,17 @@ namespace MultiplayerSample
...
+        if (m_physicsCharacter)
+        {
+            auto controller = static_cast<PhysX::CharacterController*>(m_physicsCharacter);
+            controller->SetFilterFlags(physx::PxQueryFlag::eSTATIC | physx::PxQueryFlag::eDYNAMIC | physx::PxQueryFlag::ePREFILTER);
+            if (auto callbackManager = controller->GetCallbackManager())
+            {
+                callbackManager->SetControllerFilter(CollisionLayerBasedControllerFilter);
+                callbackManager->SetObjectPreFilter(CollisionLayerBasedObjectPreFilter);
+            }
+        }
```
Comment (AMZN-puvvadar) [inline Gem/Code/Source/Components/CharacterComponent.cpp:120]: Would a client proxy have a physics character? If not, this should be fine. If so, this should probably be limited to autonomous and authority.

**Review by kberg-amzn — COMMENTED** [review]
Looking good! Consider making the changes to the prefilter, since I suspect multiple player-player collisions won't work correctly with this change as written, but otherwise this looks *far* better.. we can now control fall velocity from within ProcessInput for predictive characters which is great!

File: Gem/Code/Source/Components/CharacterComponent.cpp (hunk 2 above)
Comment (SergeyAMZN) [reply Gem/Code/Source/Components/CharacterComponent.cpp:59]: The check here is not if it's net bound or not. We can have net bound kinematic entities that will need to block the characters. 
The check here is for the kinematic rigid bodies that are dynamic on the server (i.e. they are not kinematic from the gameplay perspective). 
That's why we're specifically checking for NetworkRigidBodyRequestBus::FindFirstHandler to see if this entity is a dynamic rigid body on the server.
Good point about making components incompatible!

**Review by kberg-amzn — APPROVED** [review]
Awesome

**Pull request merged**